autograd: Add VJP and JVP rules for aten::aminmax

vijayabhaskar-ev opened the pull request:
Adds functionally correct backward (VJP) and forward (JVP) autograd rules for the aten::aminmax operator to derivatives.yaml using existing helper functions. This ensures correct eager mode differentiation.

Fixes #148808






cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @jerryzh168 @gujinghui @PenghuiCheng @jianyuh @min-jean-cho @yanbing-j @Guobing-Chen @Xia-Weiwen @snadampal @voznesenskym @penguinwu @EikanWang @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov @xmfan

vijayabhaskar-ev committed:
autograd: Add VJP and JVP rules for aten::aminmax

Adds functionally correct backward (VJP) and forward (JVP) autograd
rules for the aten::aminmax operator to derivatives.yaml using existing
helper functions. This ensures correct eager mode differentiation.

Fixes #148808

vijayabhaskar-ev commented:
@pytorchbot label "release notes: bug fix"

vijayabhaskar-ev commented:
@pytorchbot label "topic: autograd"

soulitzer commented on test/test_autograd.py:
```diff
@@ -11369,6 +11369,53 @@ def test_scatter_index_reduce_amin_amax_backprops_to_all_values(self, device):
 
             gradcheck(fn, (input, 0, idx, src, reduction), check_batched_grad=False)
 
+    def test_aminmax_backprops_to_all_values(self, device):
```
Thanks for the thorough test, it would be nice though if we could consolidate the two?

You can wrap this in a function that expose only one of the outputs and test it basically in the same way as aminmax, e.g. `amin_ = lambda x: torch.aminmax(x)[0]`

vijayabhaskar-ev replied:
I’ve updated the test case in the existing test_min_max_median_backprops_to_all_values. Let me know if any further changes are needed.

soulitzer commented:
We also have a test suite entry for aminmax that we should modify here to indicate that autograd is supported:

https://github.com/pytorch/pytorch/blame/ddfc14b3ae71dadfe0f831f6f2eff70b54d0f83f/torch/testing/_internal/common_methods_invocations.py#L14649-L14656

vijayabhaskar-ev committed:
- Updated autograd rules in FunctionsManual.cpp and FunctionsManual.h to handle amin and amax correctly.
- Modified derivatives.yaml to reflect the changes in autograd behavior.
- Updated test cases in test_autograd.py to validate the updated rules.
- Updated common_methods_invocations.py to include amin and amax in relevant test cases.

soulitzer commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -218,6 +218,47 @@ Tensor amaxamin_jvp(
+  auto grad_max_scaled = grad_max.defined()
+      ? scale_grad_by_count(
+            restore_reduced_dims(grad_max, dims_, keepdim), mask_max, dims_)
+      : at::zeros_symint(self.sym_sizes(), self.options());
```
We don't want need to materialize a zeros tensor/do an addition when if one of grad_min/grad_max are not defined, e.g. You can initialize an undefined tensor. Start with computing the grad contribution from min (or max). Then when you do the max (or min), you can check whether the tensor is still undefined and decide whether you need to add or not.

It would also be nice to do the addition in-place sometimes. You can do `a.add_(b)` if b is not a subclass. (Search for usages of `areAnyTensorSubclassLike(...)` in this file.)

soulitzer commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -218,6 +218,47 @@ Tensor amaxamin_jvp(
+    mask_max = at::ones_like(self, at::kBool);
+  } else {
+    mask_max = (max_reduced == self);
+  }
```
Why do we need special handling for nans for aminmax? Shouldn't we behave just like amax and amin, which don't have this extra handling?

vijayabhaskar-ev replied:
NaNs can affect the results differently in aminmax compared to separate amin or amax calls,  since those functions return scalars. However, I have now updated the handling of NaNs in aminmax to be consistent with how NaNs are handled  elsewhere in the codebase.

soulitzer replied:
Do you mind giving an example? I don't think amin/amax necessarily return scalars.

vijayabhaskar-ev replied:
```
# Example 1: Regular case without NaN
x = torch.tensor([1.0, 2.0, 3.0, 4.0])
print("Regular tensor:")
print(f"amin: {torch.amin(x)}")  # 1.0
print(f"amax: {torch.amax(x)}")  # 4.0 
print(f"aminmax: {torch.aminmax(x)}")  # (1.0, 4.0)

# Example 2: With NaN
x = torch.tensor([1.0, float('nan'), 3.0, 4.0])
print("\nTensor with NaN:")
print(f"amin: {torch.amin(x)}")  # nan
print(f"amax: {torch.amax(x)}")  # nan
print(f"aminmax: {torch.aminmax(x)}")  # (nan, nan)

```
When called separately, amin and amax will return scalars that are NaN if any value in the input is NaN.

However, when using aminmax with dimension reduction, we need special handling because:

- We need to track NaN values consistently across both min and max computations
- We need to preserve the relationship between corresponding min/max pairs in the output
- If all values along a reduction dimension are NaN, both min and max should be NaN
- If some values are NaN and some are not, we need consistent handling for both outputs

soulitzer replied:
Hmm I'm not sure I understand 

> When called separately, amin and amax will return scalars that are NaN if any value in the input is NaN.

amin and amax separately don't always return scalars right?

> We need to track NaN values consistently across both min and max computations

Isn't this by default what would happen if you just called amin and amax separately?

> If all values along a reduction dimension are NaN, both min and max should be NaN

If ANY value along  a reduction dimension are NaN ?

soulitzer replied:
Why do we care about what min/max do? amin/amax is different?

vijayabhaskar-ev replied:
Are you suggesting any particular solution? If so, can you specify what is the  approach  without  handling  nans for aminmax?

vijayabhaskar-ev replied:
amin and amax are single reduction operations. aminmax simultaneously returns both the minimum and maximum values.

In amin/amax, gradient is distributed to elements matching the reduction result. In aminmax, we need to handle gradient distribution for both min and max values simultaneously

vijayabhaskar-ev replied:
@soulitzer I hope the above explanation resolves your query on special handling for nans for aminmax

soulitzer reviewed (commented):
Thanks for the update, looking great. Just added a few more small comments.

vijayabhaskar-ev committed:
[autograd] Optimize tensor allocation in aminmax_backward

Makes backward pass more efficient by:
- Avoiding zeros tensor creation when only one gradient defined
- Using in-place operations where possible

vijayabhaskar-ev commented:
@IvanKobzarev Is this PR good to go? Can you review this once?

vijayabhaskar-ev commented:
@IvanKobzarev Is this PR good to go? Can you review this once?

soulitzer commented:
Yes, that is fine. Could you comment more on the NaN handling part though

vijayabhaskar-ev commented:
> Yes, that is fine. Could you comment more on the NaN handling part though

aminmax needs the NaN guard because it has two separate gradients that can land on the same nan elements. the guard splits or skips them so back prop is correct (e.g., with x=[nan,nan,nan], loss = mn + mx should yield grads 2/3 each, sum 2, but without the guard you would  double to 4 or even crash if only one output's grad is present).

`x = tensor([nan, nan, nan], requires_grad=True)` 
`mn, mx = aminmax(x) # both mn and mx are nan`

then every element of x counts as both the min and the max, because NaN ties with itself.

soulitzer commented:
Doubling NaN is still Nan though?
The gradient of amax is NaN for all elements as long as one element is NaN.

soulitzer commented:
```
`#The below is the native attempt using the  amin and amax implementation:
x = torch.tensor([float('nan'), float('nan'), float('nan')], requires_grad=True)
mn = x.min()              # NaN
mx = x.max()              # NaN
mask = (x == mn)          # empty tensor([False, False, False])

# The next line is what scale_grad_by_count does internally:
grad = 1.0 / mask.sum()   # >>> division by 0  →  inf / RuntimeError
`
```

This is expected right?
`(1.0 / mask.sum()) * mask  = inf * 0 = nan`

soulitzer commented:
> Without the guard: gradients become nan ,all gradient information is lost and NaNs propagate backward
> With the guard gradients stay finite and correctly scaled.

Yes, and the "Without the guard" case is what we want because that is what amax and amin do today

vijayabhaskar-ev committed:
• Modified the mask generation logic.
• Added a testcase to validate the new logic.

soulitzer commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -218,6 +218,43 @@ Tensor amaxamin_jvp(
+  } else if (grad_max.defined()) {
+    result = scale_grad_by_count(grad_max, max_mask, dims);
+  } else {
+    result = at::zeros_symint(self.sym_sizes(), self.options());
```
You can return an undefined Tensor() here which is semantically zero to autograd without needing to materialize the zero.
`result = Tensor();`

soulitzer commented on test/test_autograd.py:
```diff
@@ -11417,6 +11421,33 @@ def test_scatter_index_reduce_amin_amax_backprops_to_all_values(self, device):
+                fn,
+                (x,),
+                check_batched_grad=False,
+                rtol=0.6,
```
This is quite a large rtol
At this point let's just not use gradcheck
We should just reuse the `test_min_max_median_backprops_to_all_values` to test that amax,amin,aminmax  evenly distributes the gradients just like min/max

soulitzer replied:
I see, let's test only the 1/3 results part on the non-NaN inputs for amin/amax/aminmax

vijayabhaskar-ev committed:
Refactored  aminmax_backward  and added testcases for amin, amax and aminmax

vijayabhaskar-ev committed:
Update dimension parameter to use std::optional<int64_t> and refactored testcase

soulitzer commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -218,6 +218,43 @@ Tensor amaxamin_jvp(
+    result = scale_grad_by_count(grad_min, min_mask, dims);
+
+    if (grad_max.defined()) {
+      auto grad_max_result = scale_grad_by_count(grad_max, max_mask, dims);
```
Are you missing a restore_reduced_dims call on grad_max?
